src/common/com/intellij/plugins/haxe/model/type/HaxeClassReference.java

```diff
@@ -44,15 +49,29 @@ public HaxeClassReference(String name, @NotNull PsiElement elementContext) {
   }
 
   private String getClassName(HaxeClassModel clazz) {
+    Pair<Integer, String> className = clazz.haxeClass.getUserData(CLASS_NAME_KEY);
+    if (className == null || className.first != clazz.haxeClass.hashCode()) {
```

Comment:
What's the use case for the hash code?  If there is none, then why store it as a pair?  If there is a use case, why no API to retrieve it?

Reply:
Well this part is a bit tricky. For big anonymous structures we are getting lots of allocations, and this is why we need shortening, preserving and re-validating by hashCode.

Reply:
Perhaps we should only shorten _anonymous_ structure names then?